src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.cassandra.io.sstable.format.SSTableFormat;
+import org.apache.cassandra.io.sstable.format.Version;
+import org.apache.cassandra.io.sstable.format.big.BigTableBlockWriter;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.streaming.ProgressInfo;
+import org.apache.cassandra.streaming.StreamReceiver;
+import org.apache.cassandra.streaming.StreamSession;
+import org.apache.cassandra.streaming.messages.StreamMessageHeader;
+import org.apache.cassandra.utils.FBUtilities;
+
+/**
+ * CassandraBlockStreamReader reads SSTable off the wire and writes it to disk.
+ */
+public class CassandraBlockStreamReader implements IStreamReader
+{
+    private static final Logger logger = LoggerFactory.getLogger(CassandraBlockStreamReader.class);
+    private final TableId tableId;
+    private final StreamSession session;
+    private final int sstableLevel;
+    private final SerializationHeader.Component header;
+    private final int fileSeqNum;
+    private final ComponentManifest manifest;
+    private final SSTableFormat.Type format;
+    private final Version version;
+    private final DecoratedKey firstKey;
+
+    public CassandraBlockStreamReader(StreamMessageHeader header, CassandraStreamHeader streamHeader, StreamSession session)
+    {
+        if (session.getPendingRepair() != null)
+        {
+            // we should only ever be streaming pending repair
+            // sstables if the session has a pending repair id
+            if (!session.getPendingRepair().equals(header.pendingRepair))
+                throw new IllegalStateException(String.format("Stream Session & SSTable ({}) pendingRepair UUID mismatch.",
+                                                              header.tableId));
+        }
+        this.session = session;
+        this.tableId = header.tableId;
+        this.manifest = streamHeader.componentManifest;
+        this.sstableLevel = streamHeader.sstableLevel;
+        this.header = streamHeader.header;
+        this.format = streamHeader.format;
+        this.fileSeqNum = header.sequenceNumber;
+        this.version = streamHeader.version;
+        this.firstKey = streamHeader.firstKey;
+    }
+
+    /**
+     * @param inputPlus where this reads data from
+     * @return SSTable transferred
+     * @throws IOException if reading the remote sstable fails. Will throw an RTE if local write fails.
+     */
+    @SuppressWarnings("resource") // input needs to remain open, streams on top of it can't be closed
+    @Override
+    public SSTableMultiWriter read(DataInputPlus inputPlus) throws IOException
+    {
+        long totalSize = manifest.getTotalSize();
+
+        ColumnFamilyStore cfs = ColumnFamilyStore.getIfExists(tableId);
+
+        if (cfs == null)
+        {
+            // schema was dropped during streaming
+            throw new IOException("CF " + tableId + " was dropped during streaming");
```

Comment:
We generally don't call things `CF` in any new code anymore, `table` is the preferred terminology.

Reply:
Cherry picked